File: 人工智能概论/exp1/game.py
import pygame
import sys
from constants import *
from player import Player
from hook import Hook
from items import ItemManager
from levels import LevelManager
from ui import UI
from utils import load_image, load_sound

class GoldMinerGame:
    def __init__(self):
        pygame.init()
        self.screen = pygame.display.set_mode((SCREEN_WIDTH, SCREEN_HEIGHT))
        pygame.display.set_caption("黄金矿工")
        self.clock = pygame.time.Clock()
        
        # 初始化游戏组件
        self.player = Player()
        self.hook = Hook(self.player.position)
        self.item_manager = ItemManager()
        self.level_manager = LevelManager()
        self.ui = UI()
        
        # 加载资源
        self.background = load_image("resources/images/background.png", BLACK, (SCREEN_WIDTH, SCREEN_HEIGHT))
        self.sounds = {
            'throw': load_sound("resources/sounds/throw.wav"),
            'grab': load_sound("resources/sounds/grab.wav"),
            'cash': load_sound("resources/sounds/cash.wav")
        }
        
        # 游戏状态
        self.game_state = "start"  # start, playing, level_complete, game_over
        self.time_left = 0
        self.level_start_time = 0
        
    def handle_events(self):
        """处理游戏事件"""
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                return False
            
            if event.type == pygame.KEYDOWN:
                if event.key == pygame.K_ESCAPE:
                    return False
                
                if self.game_state == "start" and event.key == pygame.K_SPACE:
                    self.start_game()
                
                elif self.game_state == "playing" and event.key == pygame.K_SPACE:
                    if self.hook.state == "swinging":
                        self.hook.throw()
                        self.play_sound('throw')
                
                elif self.game_state == "level_complete" and event.key == pygame.K_SPACE:
                    self.next_level()
                
                elif self.game_state == "game_over" and event.key == pygame.K_r:
                    self.reset_game()
        
        return True
    
    def start_game(self):
        """开始游戏"""
        self.game_state = "playing"
        self.time_left = self.level_manager.get_current_time_limit()
        self.level_start_time = pygame.time.get_ticks()
        self.item_manager.generate_level_items(self.level_manager.current_level)
    
    def next_level(self):
        """进入下一关"""
        self.level_manager.advance_level()
        self.start_game()
    
    def reset_game(self):
        """重置游戏"""
        self.player = Player()
        self.hook = Hook(self.player.position)
        self.item_manager = ItemManager()
        self.level_manager.reset()
        self.game_state = "start"
    
    def update(self, dt):
        """更新游戏状态"""
        if self.game_state == "playing":
            # 更新钩子
            self.hook.update(dt)
            
            # 检查碰撞
            if self.hook.state == "throwing":
                caught_item = self.item_manager.check_collision(self.hook.get_hook_rect())
                if caught_item:
                    self.hook.catch_item(caught_item)
                    self.play_sound('grab')
            
            # 检查钩子是否回到玩家位置
            if self.hook.state == "pulling" and self.hook.length <= 0:
                if self.hook.caught_item:
                    self.player.add_score(self.hook.caught_item.value)
                    self.play_sound('cash')
                    self.item_manager.remove_caught_items()
                    self.hook.caught_item = None
            
            # 更新时间
            current_time = pygame.time.get_ticks()
            elapsed_seconds = (current_time - self.level_start_time) // 1000
            self.time_left = max(0, self.level_manager.get_current_time_limit() - elapsed_seconds)
            
            # 检查关卡完成
            if self.player.score >= self.level_manager.get_current_target():
                self.game_state = "level_complete"
            
            # 检查时间结束
            if self.time_left <= 0:
                self.game_state = "game_over"
    
    def play_sound(self, sound_name):
        """播放音效"""
        sound = self.sounds.get(sound_name)
        if sound:
            sound.play()
    
    def draw(self):
        """绘制游戏"""
        # 绘制背景
        self.screen.blit(self.background, (0, 0))
        
        if self.game_state == "start":
            self.ui.draw_start_screen(self.screen)
        
        elif self.game_state == "playing":
            # 绘制游戏元素
            self.item_manager.draw(self.screen)
            self.hook.draw(self.screen)
            self.player.draw(self.screen)
            self.ui.draw_game_ui(self.screen, self.player, self.level_manager, self.time_left)
        
        elif self.game_state == "level_complete":
            self.ui.draw_level_complete(self.screen, self.level_manager)
        
        elif self.game_state == "game_over":
            self.ui.draw_game_over(self.screen, self.player, self.level_manager)
        
        pygame.display.flip()
    
    def run(self):
        """运行游戏主循环"""
        running = True
        
        while running:
            dt = self.clock.tick(FPS) / 1000.0  # 转换为秒
            
            running = self.handle_events()
            self.update(dt)
            self.draw()
        
        pygame.quit()
        sys.exit()

File: 人工智能概论/exp1/gold-miner-v7/js/game.js
// 游戏主类
class GoldMinerGame {
    constructor(canvas) {
        this.canvas = canvas;
        this.ctx = canvas.getContext('2d');
        
        // 游戏状态
        this.gameState = GAME_STATES.MENU;
        this.player = new Player();
        this.hook = new Hook(this.player.position);
        this.itemManager = new ItemManager();
        
        // 游戏参数
        this.level = 1;
        this.timeLeft = 0;
        this.levelStartTime = 0;
        
        // 鼠标状态
        this.mouse = {
            x: CANVAS_WIDTH / 2,
            y: 100,
            isDown: false
        };
        
        // 图像资源
        this.background = null;
        
        // 初始化
        this.init();
    }
    
    async init() {
        // 加载资源
        await this.loadResources();
        
        // 设置事件监听器
        this.setupEventListeners();
        
        // 开始游戏循环
        this.gameLoop();
    }
    
    async loadResources() {
        this.background = await Utils.loadImage(
            'assets/images/background.png', 
            '#1a3c5c',
            {width: CANVAS_WIDTH, height: CANVAS_HEIGHT}
        );
        
        await this.hook.loadImage();
    }
    
    setupEventListeners() {
        // 鼠标移动事件
        this.canvas.addEventListener('mousemove', (e) => {
            const rect = this.canvas.getBoundingClientRect();
            this.mouse.x = e.clientX - rect.left;
            this.mouse.y = e.clientY - rect.top;
        });
        
        // 鼠标点击事件
        this.canvas.addEventListener('mousedown', (e) => {
            if (e.button === 0) { // 左键
                this.mouse.isDown = true;
                
                if (this.gameState === GAME_STATES.PLAYING) {
                    this.hook.throw();
                }
            }
        });
        
        this.canvas.addEventListener('mouseup', (e) => {
            if (e.button === 0) {
                this.mouse.isDown = false;
            }
        });
    }
    
    startLevel(level) {
        this.level = level;
        this.gameState = GAME_STATES.PLAYING;
        this.timeLeft = this.getCurrentLevelConfig().timeLimit;
        this.levelStartTime = Date.now();
        
        // 重置玩家分数（如果是从菜单开始）
        if (level === 1) {
            this.player.reset();
        }
        
        // 生成关卡物品
        this.itemManager.generateLevelItems(level);
        
        // 重置钩子
        this.hook.state = 'idle';
        this.hook.length = 0;
        this.hook.caughtItem = null;


        console.log("move")
        
        this.updateUI();
        
        // 隐藏所有屏幕
        this.hideAllScreens();
    }
    
    nextLevel() {
        if (this.level < LEVEL_CONFIG.length) {
            this.startLevel(this.level + 1);
        } else {
            // 所有关卡完成
            this.showGameWin();
        }
    }
    
    restartLevel() {
        this.startLevel(this.level);
    }
    
    showMenu() {
        this.gameState = GAME_STATES.MENU;
        this.showScreen('start-menu');
    }
    
    update() {
        if (this.gameState === GAME_STATES.PLAYING) {
            // 更新钩子
            this.hook.update(this.mouse);

            // console.log("hook state:", this.hook.state);
            
            // 检查碰撞
            if (this.hook.state === 'throwing') {
                const caughtItem = this.itemManager.checkHookCollision(this.hook.getCircle());
                if (caughtItem) {
                    this.hook.catchItem(caughtItem);
                }
            }
            
            // 检查钩子是否回到玩家位置
            if (this.hook.state === 'pulling' && this.hook.length <= 100) {
                console.log("hook returned to player");
                const caughtItem = this.hook.releaseItem();
                if (caughtItem) {
                    this.player.addScore(caughtItem.value);

                    console.log(`current: ${this.player.score}`);

                    this.itemManager.removeItem(caughtItem);
                    this.updateUI();
                }
            }
            
            // 更新时间
            this.updateTime();
            
            // 检查游戏状态
            this.checkGameState();
        }
    }
    
    updateTime() {
        const elapsedSeconds = Math.floor((Date.now() - this.levelStartTime) / 1000);
        this.timeLeft = Math.max(0, this.getCurrentLevelConfig().timeLimit - elapsedSeconds);
        
        // 检查时间结束
        if (this.timeLeft <= 0 && this.gameState === GAME_STATES.PLAYING) {
            this.showGameOver();
        }
        
        this.updateUI();
    }
    
    checkGameState() {
        // 检查关卡完成
        if (this.player.score >= this.getCurrentLevelConfig().targetScore) {
            this.showLevelComplete();
        }
    }
    
    getCurrentLevelConfig() {
        return LEVEL_CONFIG[this.level - 1];
    }
    
    updateUI() {
        console.log(`good1${this.player.score}`);
        document.getElementById('score').textContent = `分数: $${this.player.score}`;
        document.getElementById('level').textContent = `关卡: ${this.level}`;
        document.getElementById('target').textContent = `目标: $${this.getCurrentLevelConfig().targetScore}`;
        document.getElementById('time').textContent = `时间: ${this.timeLeft}秒`;
    }
    
    showLevelComplete() {
        this.gameState = GAME_STATES.LEVEL_COMPLETE;
        
        const levelCompleteScreen = document.getElementById('level-complete-screen');
        const levelCompleteText = document.getElementById('level-complete-text');
        
        if (this.level < LEVEL_CONFIG.length) {
            levelCompleteText.textContent = `关卡 ${this.level} 完成！`;
        } else {
            levelCompleteText.textContent = "恭喜！你已完成所有关卡！";
        }
        
        this.hideAllScreens();
        levelCompleteScreen.classList.remove('hidden');
    }
    
    showGameOver() {
        this.gameState = GAME_STATES.GAME_OVER;
        document.getElementById('final-score').textContent = `最终分数: $${this.player.score}`;
        
        this.hideAllScreens();
        document.getElementById('game-over-screen').classList.remove('hidden');
    }
    
    showGameWin() {
        this.gameState = GAME_STATES.GAME_WIN;
        document.getElementById('win-score').textContent = `最终分数: $${this.player.score}`;
        
        this.hideAllScreens();
        document.getElementById('game-win-screen').classList.remove('hidden');
    }
    
    hideAllScreens() {
        const screens = document.querySelectorAll('.screen');
        screens.forEach(screen => {
            screen.classList.add('hidden');
        });
    }
    
    draw() {
        // 清除画布
        this.ctx.clearRect(0, 0, CANVAS_WIDTH, CANVAS_HEIGHT);
        
        // 绘制背景
        if (this.background) {
            this.ctx.drawImage(this.background, 0, 0, CANVAS_WIDTH, CANVAS_HEIGHT);
        } else {
            // 默认背景
            this.ctx.fillStyle = '#1a3c5c';
            this.ctx.fillRect(0, 0, CANVAS_WIDTH, CANVAS_HEIGHT);
            
            // 绘制地面
            this.ctx.fillStyle = '#8B4513';
            this.ctx.fillRect(0, CANVAS_HEIGHT - 100, CANVAS_WIDTH, 100);
        }
        
        // 绘制物品
        this.itemManager.draw(this.ctx);
        
        // 绘制玩家
        this.player.draw(this.ctx);
        
        // 绘制钩子和瞄准线
        this.hook.draw(this.ctx);
        this.hook.drawAimLine(this.ctx, this.mouse);
        
        // 调试信息：显示碰撞区域
        if (this.gameState === GAME_STATES.PLAYING) {
            // 绘制钩子碰撞区域
            const hookCircle = this.hook.getCircle();
            this.ctx.strokeStyle = 'rgba(255, 0, 0, 0.5)';
            this.ctx.lineWidth = 1;
            this.ctx.beginPath();
            this.ctx.arc(hookCircle.x, hookCircle.y, hookCircle.radius, 0, Math.PI * 2);
            this.ctx.stroke();
            
            // 绘制物品碰撞区域
            for (const item of this.itemManager.items) {
                const itemCircle = item.getCircle();
                this.ctx.strokeStyle = 'rgba(0, 255, 0, 0.5)';
                this.ctx.lineWidth = 1;
                this.ctx.beginPath();
                this.ctx.arc(itemCircle.x, itemCircle.y, itemCircle.radius, 0, Math.PI * 2);
                this.ctx.stroke();
            }
        }
    }
    
    gameLoop() {
        this.update();
        this.draw();
        
        setTimeout(() => {
            this.gameLoop();
        }, 1000 / FPS);
    }
}